README.md
# 白洞 / 意识海

## 空间化、离散、不断膨胀的留言板

> 0.85: 程序化音乐差不多可以听了（使用采样音源，扩展Sampler支持多通道输出避免重复创建实例，重写Panner，动态调整的效果器、淡入淡出） 很多的性能优化（解决了像素画板断断续续、走着走着卡一下等问题） 重做界面（世界坐标、圆形遮罩）

### 简介

白洞 是一个空间化的留言板。你可以在里留下会说话的像素物体。它们会发出声音，然后飘走。

### 规则

1. 不沟通

2. 不关心别人

3. 保持独立、保持距离、保持自由

4. 违反规则我会引爆核弹 :fire:

### 说明

#### 1. 世界的运转方式

中心是现在，距离是时间。新创建的对象出现在世界中心，已有对象随时间推移和密度增加向外扩散。

#### 2. 对象的参数

- 形象：对象的样子。
  点击画布上方的＋可以画动画。
  默认提供 8 种颜料，创建对象有几率获得一次性的特殊颜料。

- 对话：对象说的话。每句话之间用回车分割。
  没有对话的物体对空间密度影响较小。
  （欢迎批量生产宇宙垃圾）

- 名字：默认为？？？，给对象取名字消耗便签。

- 链接：消耗箱子让对象携带一个网址（url）。

- 体积：决定对象的大小。设置为 M 以上 消耗肥料。

- 深度：
  决定对象在 z 轴的位置，近快远慢。
  设置 0.4~0.8 或 1.2~1.6 需要消耗镜片。
  可以用这个参数模拟前景和背景。

- 声音（未实装）：
  决定对象发出的音色。

#### 3. 程序化音乐（beta）

背景音乐是根据空间里的对象生成的。

旋律对象（深度 <= 1）：有不同的音高、音色、时值、位置，每若干小节重复。

和声对象（深度 > 1）：有不同的音域、音色。根据每天随机生成的和弦进行演奏和弦音。

对象的音色和音高会受对象的参数影响。

### 平台/版本

- 网页版：[地址 1](https://lcl.yu-me.workers.dev) [地址 2](https://lcl-web.herokuapp.com)
- 将网页安装为 PWA（网页应用）
  - Chrome 等浏览器会自动弹出安装提示。你也可以点击本页面顶部的 “安装到桌面” 按钮安装。
  - IOS 在 Safari 中点击 `分享 - 添加到主屏幕` 可以将本应用安装到主屏。
- 安卓版（开发中）
- PC 版（开发中）

- [Github 仓库](https://github.com/john-walks-slow/lcl)
- [报告 BUG / 提建议](https://github.com/john-walks-slow/lcl/issues/new)

### 特别感谢

本项目的像素画板是基于[jvalen/pixel-art-react](https://github.com/jvalen/pixel-art-react)修改的

(MIT Copyright © 2016 Javier Valencia Romero)

### 待办事项

##### 当前版本
- [ ] :package: PC 版开发
- [ ] :package: 安卓版开发
- [ ] :sparkles: i18n
- [ ] :sparkles: 自动翻译
- [ ] :books: 

##### 今后
- [ ] :shirt: SEO

### Changelog

##### [详细版：https://github.com/john-walks-slow/lcl/blob/master/history.md](https://github.com/john-walks-slow/lcl/blob/master/history.md)
- 0.85: 程序化音乐差不多可以听了（使用采样音源，扩展Sampler支持多通道输出避免重复创建实例，重写Panner，动态调整的效果器、淡入淡出） 很多的性能优化（解决了像素画板断断续续、走着走着卡一下等问题） 重做界面（世界坐标、圆形遮罩）
- 0.8: 初步完成程序化音乐 重做 ui 设置 M 以下及一级深度不消耗道具 离线可用 重构代码 解决很多 bug
- 0.7: 新增道具功能 新增超链接功能 新增 PWA 支持 优化缓存机制 重做加载界面 优化数值
- 0.6: 迁移到 MongoDB 基本功能可用

### 我为什么做这个？

有一天我玩了很久游戏，洗碗的时候我在想我是不是可以做游戏。

如果做游戏，我一定要在游戏里留下一点自己的痕迹，我也希望玩家可以在游戏里留下他们的痕迹，那不如创造一个大家可以随便留下痕迹的地方。

我对这里的设想是一个“共享的梦空间”，把大家脑袋里的想象放在同一个世界里会不会有趣？


history.md
- 0.85: 程序化音乐差不多可以听了（使用采样音源，扩展Sampler支持多通道输出避免重复创建实例，重写Panner，动态调整的效果器、淡入淡出） 很多的性能优化（解决了像素画板断断续续、走着走着卡一下等问题） 重做界面（世界坐标、圆形遮罩）
  - [x] :alien: uqn上发帖
  - [x] :ambulance: jsdelivr 出问题了，增加 cdn fallback
  - [x] :art: 调整 UI 色值，与色板吻合
  - [x] :art: 新图标
  - [x] :art: 修复字体缺失符号，增加日语、英语、繁体字
  - [x] :art: 重新设计 Add Page
  - [x] :art: 重新设计 pixel editor，布局更合理，大屏下工具栏横置，简化画布操作
  - [x] :bug: 创建 game 前未更新屏幕坐标
  - [x] :bug: 扩大 canvas/选择颜料时按钮位移
  - [x] :bug: 连续点击 map 相机卡死
  - [x] :bug: Add Page 改变屏幕大小导致坐标偏移
  - [x] :bug: canvas 和 preview 像素间白边
  - [x] :bug: frames drag&drop / scrollbar 重叠
  - [x] :bug: Inventory reducer 未及时更新
  - [x] :fire: 拒绝“白洞”这个名字，白洞是一个只发射，不吸收的时空区域。白洞更真，意识海更有指向性。
  - [x] :fire: 字体加载超级慢，更换字体
  - [x] :musical_note: 尝试不同音色
  - [x] :musical_note: 重写panner算法（玩家在对象中的情况）
  - [x] :musical_note: 根据对象参数动态调整的混响、延迟、eq
  - [x] :musical_note: 和弦和旋律的时值更有随机性
  - [x] :musical_note: 解决很多bug，调整参数
  - [x] :musical_note: 每天会随机生成调性
  - [x] :musical_note: 完成 pannernode 替代方案
  - [x] :musical_note: 增加混响和延迟
  - [x] :musical_note: 重构，大量性能优化，动态挂载 synth，仅更新可听见的 node，新手机上基本不卡
  - [x] :musical_note: 重写sampler
  - [x] :musical_note: start time 报错
  - [x] :rocket: 细分onmousemove，解决像素画板断断续续的问题
  - [x] :rocket: 研究到底性能的瓶颈在哪里
  - [x] :rocket: object pooling
  - [x] :sparkles: 尝试粒子（否决，用户可以自行创建粒子）
  - [x] :sparkles: 分辨率自适应，优化 item 大小
  - [x] :sparkles: 恒定的坐标系统
  - [x] :sparkles: 没有对话的对象不增加空间密度，会更快飘走，不会生成道具和颜料
  - [x] :sparkles: 深度不为 1 的对象不会挡路
  - [x] :sparkles: 现在只有视野范围内的对象会根据 zfactor 更新坐标
  - [x] :sparkles: 重新设计颜料色板，默认赠送 8 色，奖励颜色从 64 色中抽取
  - [x] :sparkles: undo redo 现在会以一次完整的操作为单位
  - [x] :wrench: 调试参数
  - [x] :zap: 扩大按钮面积
- 0.8: 初步完成程序化音乐 解决很多 bug 重做 ui 设置 M 以下及一级深度不消耗道具 离线可用 重构代码
  - [x] :art: 调整响应式布局
  - [x] :art: 重做滤镜
  - [x] :art: 重做 ui
  - [x] :books: 优化 README
  - [x] :bug: 背景/前景 pushable
  - [x] :bug: 检查 player scheme
  - [x] :bug: 输入事件捕获异常
  - [x] :bug: 新对象卡启动
  - [x] :bug: 再创建一个时页面重复刷新
  - [x] :bug: item dialog 窗口位置
  - [x] :bug: link dialog 按钮失效
  - [x] :bug: link dialog 隐藏虚拟键盘
  - [x] :bug: resize 时更新 camera viewport
  - [x] :fire: 重构 Configurations
  - [x] :fire: 重构 GenerativeMusic
  - [x] :fire: 重构 ObjectData
  - [x] :fire: 重构 ObjectGroup
  - [x] :musical_note: 音乐能听
  - [x] :sparkles: 道具生成率降低，道具大小减少
  - [x] :sparkles: 发言前检测用户是否阅读本文档
  - [x] :sparkles: 更好的报错界面
  - [x] :sparkles: 键盘快捷键
  - [x] :sparkles: 降低道具重要性，深度-2~2 不消耗镜片
  - [x] :sparkles: 配置 sw 离线可用
  - [x] :sparkles: 优化自动转屏逻辑
  - [x] :sparkles: 重写渲染算法（pixel perfect）
  - [x] :sparkles: Generative Music
  - [x] :sparkles: ui 可以隐藏
  - [x] :wrench: 调整数值
  - [x] :wrench: 设置 M 以下不需要肥料
  - [x] :wrench: 使用 history api 优化后退键
  - [x] :wrench: 提高服务器安全性
  - [x] :wrench: ui 事件优化


public/script.js

const WINDOW_W = window.innerWidth || document.body.clientWidth;
const WINDOW_H = window.innerHeight || document.body.clientHeight;
const WINDOW_CENTER_X = WINDOW_W / 2;
const WINDOW_CENTER_Y = WINDOW_H / 2;
const PLAYER_TARGET_W = Math.min(WINDOW_W / 11, WINDOW_H / 11 / 44 * 37);
const PLAYER_TARGET_H = Math.min(WINDOW_W / 11, WINDOW_H / 11 / 44 * 37) / 37 * 44;
// const PLAYER_TARGET_W = Math.min(WINDOW_W / 15, WINDOW_H / 15 / 44 * 37);
// const PLAYER_TARGET_H = Math.min(WINDOW_W / 15, WINDOW_H / 15 / 44 * 37) / 37 * 44;
const OBJECT_W = { XL: PLAYER_TARGET_H * 1.8, L: PLAYER_TARGET_H * 1.3, M: PLAYER_TARGET_H * 0.9, S: PLAYER_TARGET_H * 0.7, XS: PLAYER_TARGET_H * 0.4 }
// density: target object count in a circle (r=OBJECT_M_W*8)
const DIALOG_HEIGHT = WINDOW_H / 3.5;
const TEXT_PADDING_W = Math.min(WINDOW_W / 15, 50);
const TEXT_PADDING_H = Math.min(WINDOW_H / 25);
const PADDING_BETWEEN = 10;
const DIALOG_PADDING_W = Math.min(WINDOW_W / 20, 70);
const DIALOG_PADDING_H = Math.min(WINDOW_W / 25, 30);
const FONT_SIZE = Math.max(WINDOW_H / 30, WINDOW_W / 40);
const FONT_SIZE_HEADER = FONT_SIZE * 1.2;
// const DIALOG_HEIGHT = FONT_SIZE*2+FONT_SIZE_HEADER+TEXT_PADDING_H*3 +PADDING_BETWEEN;
const FONT_FAMILY = "pixel"
const FONT_FAMILY_HEADER = "pixel"
// const TIME_DELAY = 60 * 60 * 1000;
const TIME_DELAY = 0;
const RANDOM_ZONE_W = OBJECT_W.XL;
var DENSITY_OFFSET = OBJECT_W.M * 2;
const MOVE_SPEED = PLAYER_TARGET_W * 1.1;
const ZONE_SIZE = MOVE_SPEED * 30;
const timestamp = Date.parse(new Date());


function createTestObject(object) {
  object._id = 3;
  object.seed = [Math.random() * 360, Math.random() ** 0.5];
  return object;
}

function vectorAngle(x, y) {
  let mX = Math.sqrt(x.reduce((acc, n) => acc + Math.pow(n, 2), 0));
  let mY = Math.sqrt(y.reduce((acc, n) => acc + Math.pow(n, 2), 0));
  return Math.acos(x.reduce((acc, n, i) => acc + n * y[i], 0) / (mX * mY));
};

class Dialog extends Phaser.GameObjects.Container {
  constructor(scene) {
    super(scene);
    this.scene = scene;
    this.depth = 999;
    this.dialogWindow = this.scene.add.sprite(WINDOW_CENTER_X, WINDOW_H - DIALOG_HEIGHT / 2 - DIALOG_PADDING_H, "dialog").setDisplaySize(WINDOW_W - DIALOG_PADDING_W, DIALOG_HEIGHT);
    this.dialogHeader = this.scene.add.text(DIALOG_PADDING_W + TEXT_PADDING_W, WINDOW_H - DIALOG_HEIGHT - DIALOG_PADDING_H + TEXT_PADDING_H, "???",
      {
        color: 0xFFFFFF,
        fontStyle: "bold",
        // fixedWidth: WINDOW_W - DIALOG_PADDING_W - TEXT_PADDING*2,
        // fixedHeight: DIALOG_HEIGHT - TEXT_PADDING*2,
        fontFamily: FONT_FAMILY_HEADER,
        fontSize: (FONT_SIZE_HEADER).toString() + "px",
      });
    this.dialogText = this.scene.add.text(DIALOG_PADDING_W + TEXT_PADDING_W, WINDOW_H - DIALOG_HEIGHT - DIALOG_PADDING_H + TEXT_PADDING_H + FONT_SIZE_HEADER + PADDING_BETWEEN, "",
      {
        color: 0xFFFFFF,
        // fixedWidth: WINDOW_W - DIALOG_PADDING_W - TEXT_PADDING*2,
        // fixedHeight: DIALOG_HEIGHT - TEXT_PADDING*2,
        fontFamily: FONT_FAMILY,
        fontSize: FONT_SIZE.toString() + "px",
        wordWrap: { width: WINDOW_W - DIALOG_PADDING_W * 2 - TEXT_PADDING_W * 2 + 10, useAdvancedWrap: true }
      });
    this.add([this.dialogWindow, this.dialogText, this.dialogHeader]);

    this.setScrollFactor(0);
    this.setAlpha(0);
    this.dialogIndex = 0;
    this.inDialog = false;
    this.dialogFadeIn = this.scene.tweens.create({
      targets: this,
      alpha: { from: 0, to: 1 },
      ease: 'Linear',       // 'Cubic', 'Elastic', 'Bounce', 'Back'
      duration: 500,
    });
    this.dialogFadeOut = this.scene.tweens.create({
      targets: this,
      alpha: { from: 1, to: 0 },
      ease: 'Linear',       // 'Cubic', 'Elastic', 'Bounce', 'Back'
      duration: 500,
    });
    this.scene.add.existing(this);
  }
  proceedDialog() {
    if (this.inDialog) {
      this.dialogIndex++;
      if (this.dialogIndex > this.sentences.length - 1) {
        this.dialogFadeOut.play().on("complete", () => { this.inDialog = false; });
        // dialogWindow.off('pointerdown');
        this.scene.input.off('pointerdown');
        this.scene.input.keyboard.off('keydown-SPACE');
      } else {
        this.dialogText.setText(this.sentences[this.dialogIndex]);
      }
    }
  }
  showDialog(dialog, name) {
    this.inDialog = true;
    // dialogWindow.on('pointerdown', () => { this.proceedDialog() });
    this.scene.input.on('pointerdown', () => { this.proceedDialog() });
    this.scene.input.keyboard.on('keydown-SPACE', () => { this.proceedDialog() });

    this.dialogFadeIn.play();
    this.sentences = dialog;
    this.dialogIndex = 0;
    if (name != "") { this.dialogHeader.setText(name) };
    this.dialogText.setText(this.sentences[this.dialogIndex]);
    console.log(this.sentences);
  }
}

class LoadingScene extends Phaser.Scene {
  constructor() {
    super({
      key: 'LoadingScene',
    })
  }
  preload() { }

  create() {
    var app = feathers();
    // Connect to a different URL
    var restClient = feathers.rest();
    // Configure an AJAX library (see below) with that client
    app.configure(restClient.fetch(window.fetch));
    app.service('objects').find().then(({ data }) => {
      let objectList = data;
      console.log(objectList);
      this.load.image('bg', 'assets/bg.png');
      this.load.image('dialog', 'assets/dialog.png');
      this.load.spritesheet('player', 'assets/White.png', { frameWidth: 37, frameHeight: 44 });
      // [...Array(10000)].forEach(() => {
      //   objectList.push(createTestObject({
      //     "birthday": 1637818994985,
      //     "movement": "static",
      //     "size": "L",
      //     "columns": 18,
      //     "rows": 18,
      //     "zFactor": 1,
      //     "isAnimate": false,
      //     "dialog": "我是量产机1号。\n量产机只有我。\n我曾经统治过世界！\n你不信？我可是作者钦定的。你没发现我长得和你很像吗？\n写代码不容易，一直在查来查去解决琐碎的问题，会有积累吗？".split("\n"),
      //     "link": null
      //   }))
      // });
      objectList.forEach((o) => {
        if (timestamp - o.birthday < (48 * 60 * 60)) { DENSITY_OFFSET -= 10; }
        switch (o.isAnimate) {
          case true:
            var shardsImg = new Image();
            shardsImg.onload = () => {
              this.textures.addSpriteSheet("object" + o._id, shardsImg, { frameWidth: o.columns, frameHeight: o.rows });
            };
            shardsImg.src = o.blobURI;
            // this.load.spritesheet("object" + o._id, 'assets/objects/' + o._id + '.png', { frameWidth: o.columns, frameHeight: o.rows });
            break;
          default:
            this.textures.addBase64("object" + o._id, o.blobURI);
            // this.load.image("object" + o._id, 'assets/objects/' + o._id + '.png')
            break;
        }
      });
      objectList.sort((a, b) => b.birthday - a.birthday)
      DENSITY_OFFSET = Math.min(OBJECT_W.L, DENSITY_OFFSET);
      this.load.on("complete", () => {
        console.log(objectList);
        this.scene.start("MainScene", { "objectList": objectList });
        this.scene.stop("LoadingScene");
      }, this);
      this.load.on('progress', (progress) => {
        console.log(progress);
        this.label.text = 'Now Loading ' + ".".repeat(Math.round(progress * 15));
      })
      this.load.start();
    })
    this.add.rectangle(WINDOW_CENTER_X, WINDOW_CENTER_Y, WINDOW_W, WINDOW_H, 0xFFFFFF);
    this.label = this.add.text(30, WINDOW_CENTER_Y, 'Now Loading ', { align: "center", color: "#000000", fontSize: WINDOW_W / 20 })
  }
  update() {

  }
}

class MainScene extends Phaser.Scene {
  constructor() {
    super({
      key: 'MainScene', active: false
    });
    this.player;
    this.objects;
    this.cursors;
    this.startPosX = 0;
    this.startPosY = 0;
    this.camera;
    this.gameDialog;
    this.pointerOnPlayer = false;
  }

  preload() {
    // this.load.json('objects', 'data/objects.json');
  }
  init(data) {
    this.objectList = data.objectList;
  }
  create() {
    this.cursors = this.input.keyboard.createCursorKeys();
    this.camera = this.cameras.main;
    // this.add.tileSprite(WINDOW_CENTER_X, WINDOW_CENTER_Y,WINDOW_W,WINDOW_H,'bg');
    // this.add.rectangle(WINDOW_CENTER_X, WINDOW_CENTER_Y, WINDOW_W, WINDOW_H, 0xFFFFFF);
    this.player = this.physics.add.sprite(this.startPosX, this.startPosY, 'player')
      .setDisplaySize(PLAYER_TARGET_W, PLAYER_TARGET_H)
      .refreshBody();
    this.player.setInteractive();
    this.player.on('pointerdown', () => {
      if (this.gameDialog.inDialog) { return };
      if (this.camera.zoom == 1) { this.camera.zoomOutAnim.play(); }
      if (this.camera.zoom == this.camera.zoomOutLevel) { this.camera.zoomInAnim.play(); }
    })
    this.player.on('pointerover', () => { this.pointerOnPlayer = true })
    this.player.on('pointerout', () => { this.pointerOnPlayer = false })

    this.camera.startFollow(this.player);
    this.camera.setBackgroundColor(0xFFFFFF);
    this.camera.setFollowOffset(0, -50);
    this.camera.zoomOutLevel = 0.5;
    this.camera.zoomOutAnim = this.tweens.create({
      targets: this.camera,
      props: { 'zoom': this.camera.zoomOutLevel },
      ease: 'Cubic',       // 'Cubic', 'Elastic', 'Bounce', 'Back'
      duration: 1000,
    });
    this.camera.zoomInAnim = this.tweens.create({
      targets: this.camera,
      props: { 'zoom': 1 },
      ease: 'Cubic',       // 'Cubic', 'Elastic', 'Bounce', 'Back'
      duration: 1000,
    });
    console.log(PLAYER_TARGET_H);
    this.objectMap = [];
    this.objects = this.physics.add.group();
    let previousDate = timestamp;
    let offset;
    let dateOffset = 0;
    this.objectList.forEach((o, i) => {
      if (timestamp - o.birthday < TIME_DELAY) { return; }
      dateOffset += Math.min(1, (previousDate - o.birthday) / (14 * 24 * 60 * 60)) * 5 * PLAYER_TARGET_W;
      previousDate = o.birthday;
      offset = (DENSITY_OFFSET * (i ** 0.5));
      // console.log({ dateOffset, offset });
      // console.log(Math.min(1, (previousDate - o.birthday) / (30 * 24 * 60 * 60)));
      let rad = o.seed[0] * (Math.PI / 180);
      let sizeOffset = (PLAYER_TARGET_H + OBJECT_W[o.size] / o.zFactor) / 2;
      let distance = o.seed[1] * RANDOM_ZONE_W + offset + dateOffset + sizeOffset;
      o.x = Math.cos(rad) * distance;
      o.y = Math.sin(rad) * distance;
      o.zFactor = (o.zFactor == 1) ? o.zFactor - 0.1 + Math.random() * 0.2 : o.zFactor;
      o.ratio = o.rows / o.columns;
      o.zone = [Math.ceil(o.x / ZONE_SIZE), Math.ceil(o.y / ZONE_SIZE)]
      o.isAnimate ? this.anims.create({
        key: 'spritesheet' + o._id,
        frames: 'object' + o._id,
        frameRate: 2,
        repeat: -1,
        repeatDelay: 0
      }) : false;
      if (!this.objectMap[o.zone[0]]) {
        this.objectMap[o.zone[0]] = []
      }
      if (!this.objectMap[o.zone[0]][o.zone[1]]) {
        this.objectMap[o.zone[0]][o.zone[1]] = []
      }
      this.objectMap[o.zone[0]][o.zone[1]].push(o);
    })
    this.objectMap.getZone = (player) => {
      return [Math.ceil(player.x / ZONE_SIZE), Math.ceil(player.y / ZONE_SIZE)]
    }
    this.objectMap.getNearBy = (zone) => {
      if (!zone) { return [] }
      let results = [];
      [-1, 0, 1].forEach((v1) => {
        [-1, 0, 1].forEach((v2) => {
          results.push([zone[0] + v1, zone[1] + v2])
        })
      })
      return results
    }
    this.previousZone = null;

    this.gameDialog = new Dialog(this);
    this.colliderHandler = (o1, o2) => {
      if (this.gameDialog.inDialog) { return; }
      let currentObj;
      if (o1.texture.key == "player") {
        currentObj = o2;
      }
      else {
        currentObj = o1;
      }
      console.log(currentObj);
      let xOverlap = this.player.displayWidth / 2 + currentObj.displayWidth / 2 - Math.abs(this.player.x - currentObj.x)
      let yOverlap = this.player.displayHeight / 2 + currentObj.displayHeight / 2 - Math.abs(this.player.y - currentObj.y)
      // console.log(xOverlap, yOverlap);
      if (xOverlap < 0 || yOverlap < 0) {
        // return;
      }
      if (xOverlap < yOverlap && this.player.x < currentObj.x) {
        this.player.setX(this.player.x + -0.1);
        // console.log('left');
      }
      if (xOverlap < yOverlap && this.player.x > currentObj.x) {
        this.player.setX(this.player.x - -0.1);
        // console.log('right');
      }
      if (xOverlap > yOverlap && this.player.y < currentObj.y) {
        this.player.setY(this.player.y + -0.1);
        // console.log('up');
      }
      if (xOverlap > yOverlap && this.player.y > currentObj.y) {
        this.player.setY(this.player.y - -0.1);
        // console.log('down');
      }
      // console.log(currentObj);
      // console.log(dialog);
      if (currentObj.data.values.dialog.length > 0) { this.gameDialog.showDialog(currentObj.data.values.dialog, currentObj.data.values.name) }
      // }
    }

    this.anims.create({
      key: 'runLeft',
      frames: this.anims.generateFrameNumbers('player', { start: 8, end: 9 }),
      frameRate: 4,
      repeat: -1,
      repeatDelay: 0
    });
    this.anims.create({
      key: 'runRight',
      frames: this.anims.generateFrameNumbers('player', { start: 6, end: 7 }),
      frameRate: 4,
      repeat: -1,
      repeatDelay: 0
    });
    this.anims.create({
      key: 'runUp',
      frames: this.anims.generateFrameNumbers('player', { start: 3, end: 4 }),
      frameRate: 4,
      repeat: -1,
      repeatDelay: 0
    });
    this.anims.create({
      key: 'runDown',
      frames: this.anims.generateFrameNumbers('player', { start: 0, end: 1 }),
      frameRate: 4,
      repeat: -1,
      repeatDelay: 0
    });
    this.anims.create({
      key: 'standLeft',
      frames: [{ key: 'player', frame: 9 }],
      frameRate: 20,
      repeat: -1,
      repeatDelay: 0
    });

    this.anims.create({
      key: 'standRight',
      frames: [{ key: 'player', frame: 7 }],
      frameRate: 20,
      repeat: -1,
      repeatDelay: 0
    });
    this.anims.create({
      key: 'standUp',
      frames: [{ key: 'player', frame: 5 }],
      frameRate: 20,
      repeat: -1,
      repeatDelay: 0
    });
    this.anims.create({
      key: 'standDown',
      frames: [{ key: 'player', frame: 2 }],
      frameRate: 20,
      repeat: -1,
      repeatDelay: 0
    });


    this.player.anims.play('standRight');




  }

  update() {
    // console.log(this.player.x,this.player.y);
    let stopMovement = () => {
      if (this.player.anims.getName() == "runLeft") {
        // player.anims.stopOnFrame(player.anims.currentAnim.frames[1]);
        // player.anims.chain();
        this.player.anims.play('standLeft', true);
      }
      if (this.player.anims.getName() == "runRight") {
        // player.anims.stopOnFrame(player.anims.currentAnim.frames[1]);
        // player.anims.chain();
        // player.anims.chain('standRight');
        this.player.anims.play('standRight', true);

      }
      if (this.player.anims.getName() == "runUp") {
        // player.anims.stopOnFrame(player.anims.currentAnim.frames[1]);
        // player.anims.chain();
        this.player.anims.play('standUp', true);
      }
      if (this.player.anims.getName() == "runDown") {
        // player.anims.stopOnFrame(player.anims.currentAnim.frames[1]);
        // player.anims.chain();
        // player.anims.chain('standDown');
        this.player.anims.play('standDown', true);

      }
      moveX(0);
      moveY(0);
    }
    let moveX = (x) => {
      this.player.setVelocityX(x);
      this.objects.children.each((o) => { o.setVelocityX(- x * (o.data.values.zFactor - 1)) })
    }
    let moveY = (y) => {
      this.player.setVelocityY(y);
      this.objects.children.each((o) => { o.setVelocityY(- y * (o.data.values.zFactor - 1)) })
    }

    let updateObjects = (previousZone, currentZone) => {
      let previousZones = this.objectMap.getNearBy(previousZone);
      let currentZones = this.objectMap.getNearBy(currentZone);
      let createZones = currentZones.filter(x => !previousZones.toString().includes(x.toString()))
      let destroyZones = previousZones.filter(x => !currentZones.toString().includes(x.toString()))
      console.log({ prev: previousZones, cur: currentZones });
      console.log({ create: createZones, destroy: destroyZones });
      createZones.forEach((zone) => {
        // console.log(this.objectMap);
        if (!this.objectMap[zone[0]]) { return; }
        if (!this.objectMap[zone[0]][[zone[1]]]) { return; }
        let os = this.objectMap[zone[0]][zone[1]];
        os.forEach((o) => {
          console.log(o);
          o.instance = this.objects.create(o.x, o.y, "object" + o._id);
          // o.instance = this.physics.add.sprite(o.x,o.y,"object"+o.id);
          // console.log(this.objects);
          o.instance.depth = o.zFactor;
          o.instance.setData("name", o.name);
          o.instance.setData("dialog", o.dialog);
          o.instance.setData("zFactor", o.zFactor);
          o.ratio < 1 ?
            o.instance.setDisplaySize(OBJECT_W[o.size] / o.zFactor, OBJECT_W[o.size] * o.ratio / o.zFactor)
            : o.instance.setDisplaySize(OBJECT_W[o.size] / o.zFactor / o.ratio, OBJECT_W[o.size] / o.zFactor);
          if (o.isAnimate) {
            o.instance.anims.play('spritesheet' + o._id)
          }
          o.instance.refreshBody();
          // this.objects.add(o.instance);
          this.physics.add.collider(this.player, o.instance, this.colliderHandler);
        })
      })
      destroyZones.forEach((zone) => {
        if (!this.objectMap[zone[0]]) { return; }
        if (!this.objectMap[zone[0]][[zone[1]]]) { return; }
        let os = this.objectMap[zone[0]][zone[1]];
        os.forEach((o) => {
          this.objects.remove(o.instance, true, true)
        })
      })
    }

    if (this.gameDialog.inDialog) {
      stopMovement();
    }
    else {
      let currentZone = this.objectMap.getZone(this.player)
      // console.log(currentZone,this.previousZone);
      if (!(this.previousZone && currentZone[0] == this.previousZone[0] && currentZone[1] == this.previousZone[1])) {
        updateObjects(this.previousZone, currentZone);
        this.previousZone = currentZone;
      }
      let mousePosX = this.input.activePointer.x - WINDOW_CENTER_X;
      let mousePosY = WINDOW_CENTER_Y - this.input.activePointer.y;
      // if (this.input.activePointer.isDown && (Math.abs(mousePosX) <= PLAYER_TARGET_W * 0.5 || Math.abs(mousePosY) <= PLAYER_TARGET_H * 0.5)){
      // 	camera.toggleZoom();
      // };
      let isMouseMovement = this.input.activePointer.isDown && !this.pointerOnPlayer;
      let mouseAngle = isMouseMovement && vectorAngle([0, 1], [mousePosX, mousePosY]);
      if (this.cursors.left.isDown || this.cursors.right.isDown || this.cursors.up.isDown || this.cursors.down.isDown || isMouseMovement) {
        // camera.setZoom(1);
        this.camera.zoom == this.camera.zoomOutLevel && this.camera.zoomInAnim.play();
      }
      if ((this.cursors.left.isDown && this.cursors.up.isDown) || (isMouseMovement && mousePosX <= 0 && mouseAngle >= Math.PI * 0.125 && mouseAngle <= Math.PI * 0.375)) {
        moveX(- MOVE_SPEED * 0.75);
        moveY(- MOVE_SPEED * 0.75);
        this.gameDialog.dialogRetrigger = true;
        this.player.anims.play('runLeft', true);
      }
      else if ((this.cursors.left.isDown && this.cursors.down.isDown) || (isMouseMovement && mousePosX <= 0 && mouseAngle >= Math.PI * 0.625 && mouseAngle <= Math.PI * 0.875)) {
        moveX(- MOVE_SPEED * 0.75);
        moveY(MOVE_SPEED * 0.75);
        this.gameDialog.dialogRetrigger = true;
        this.player.anims.play('runLeft', true);
      }
      else if ((this.cursors.right.isDown && this.cursors.up.isDown) || (isMouseMovement && mousePosX > 0 && mouseAngle >= Math.PI * 0.125 && mouseAngle <= Math.PI * 0.375)) {
        moveX(MOVE_SPEED * 0.75);
        moveY(- MOVE_SPEED * 0.75);
        this.gameDialog.dialogRetrigger = true;
        this.player.anims.play('runRight', true);
      }
      else if ((this.cursors.right.isDown && this.cursors.down.isDown) || (isMouseMovement && mousePosX > 0 && mouseAngle >= Math.PI * 0.625 && mouseAngle <= Math.PI * 0.875)) {
        moveX(MOVE_SPEED * 0.75);
        moveY(MOVE_SPEED * 0.75);
        this.gameDialog.dialogRetrigger = true;
        this.player.anims.play('runRight', true);
      }
      else if ((this.cursors.left.isDown) || (isMouseMovement && mousePosX < 0 && mouseAngle >= Math.PI * 0.375 && mouseAngle <= Math.PI * 0.625)) {
        moveX(- MOVE_SPEED);
        moveY(0);
        this.gameDialog.dialogRetrigger = true;
        this.player.anims.play('runLeft', true);
      }
      else if ((this.cursors.right.isDown) || (isMouseMovement && mousePosX > 0 && mouseAngle >= Math.PI * 0.375 && mouseAngle <= Math.PI * 0.625)) {
        moveX(MOVE_SPEED);
        moveY(0);
        this.gameDialog.dialogRetrigger = true;
        this.player.anims.play('runRight', true);
      }

      else if ((this.cursors.down.isDown) || (isMouseMovement && mouseAngle >= Math.PI * 0.875)) {
        moveX(0);
        moveY(MOVE_SPEED);
        this.gameDialog.dialogRetrigger = true;
        this.player.anims.play('runDown', true);
      }

      else if ((this.cursors.up.isDown) || (isMouseMovement && mouseAngle <= Math.PI * 0.125)) {
        moveX(0);
        moveY(- MOVE_SPEED);
        this.gameDialog.dialogRetrigger = true;
        this.player.anims.play('runUp', true);
      }
      else { stopMovement(); }
    }

  }
}

var config = {
  type: Phaser.AUTO,
  width: WINDOW_W,
  height: WINDOW_H,
  physics: {
    default: 'arcade',
    arcade: {
      gravity: { y: 0 },
      debug: false
    }
  },
  scene: [LoadingScene, MainScene],
  pixelArt: true,
};

var game = new Phaser.Game(config);
